### ISSUE TITLE
alpaka_add_executable in CMAKE

### ISSUE THREAD

**Issue by erikzenker** [ComputationalRadiationPhysics/alpaka#57]
To prevent the alpaka user to treat CUDA accelerators in a special way within the CMakeList.txt it is necessary to provide a alpaka_add_executable function inside the alpaka CMAKE module.

**Comment by ax3l** [comment]
I might also want to create a shared library (aka `cuda_add_library` / `alpaka_add_library`), e.g., for [python bindings](https://github.com/ax3l/picongpu/tree/topic-pythonBindings).

**Comment by erikzenker** [comment]
Thank you for the feature :+1:  Like we were discussing in person,  the part that includes .cu files is still missing. I have an idea about that issue. What about using only .cpp files. In the case of a cuda acc, cmake creates .cu files in some temp folder which include the corresponding .cpp file. These created .cu files are then feed into the cuda_add_executable function. 

I would insert this functionality [here](https://github.com/ComputationalRadiationPhysics/alpaka/blob/develop/cmake/addExecutable.cmake#L22). What thinks our cmake experts @slizzered  and @psychocoderHPC  about that ?

**Comment by BenjaminW3** [comment]
Yes, this is a good idea. I think auto-generating ```.cu``` files for all ```.cpp``` files would be a convenient solution.

The alpaka user only writes ```.cpp``` files and calls ```ALPAKA_ADD_EXECUTABLE``` with all the source files of his project (```.cpp```, ```.cxx```, ```.cc```, ```.hpp```, ```.h```, ...). 
* all the source files have to be filtered out (NOTE: For MSVC the non source files have to be passed in to generate good project files)
* corresponding ```.cu``` files have to be generated only containing an ```#include "PATH/TO/SOURCE/FILE.extension"```. **Where to do this corectly? When to delete them? Regenerate them each time?**
* the paths to the original source files have to be replaced by the ```.cu``` wrapper files
* the paths can be passed to ```CUDA_ADD_EXECUTABLE```

Even if the CUDA accelerator is not used and there is no CUDA code, this should work correctly.
For all other accelerators the files do not have to be processed.

**Comment by ax3l** [comment]
what kind of source code transformation (example) do you mean with

> corresponding .cu files have to be generated only containing an #include "PATH/TO/SOURCE/FILE.extension". Where to do this corectly? When to delete them? Regenerate them each time?

cmake has functionality like this, e..g, from the classic configure scripts but also supports manual text-fiddeling.

Nevertheless, manual code transformation is always cumbersome. as you listed it already, it needs many definitions and puts a lot of restrictions on both the written code and the cumbersome text replacement scripts.

**Comment by ax3l** [comment]
seems like the examples [here](https://github.com/parallel-forall/code-samples/tree/master/posts/separate-compilation-linking) described in the [separable compilation](http://devblogs.nvidia.com/parallelforall/separate-compilation-linking-cuda-device-code/) blog post do not require `.cu` files. Am I missing something?

**Comment by slizzered** [comment]
Oh It seems we missed the `nvcc -x cu` option. It would work...

```
--x {c|c++|cu}                             (-x)                              
        Explicitly specify the language for the input files, rather than letting
        the compiler choose a default based on the file name suffix.
        Allowed values for this option:  'c','c++','cu'.
```

Apart from that, the automatic generation of `.cu`-files should not be overly complicated, since the files always contain nothing but a single include to their original `.cpp`-file.

**Comment by ax3l** [comment]
oh `--x` seems reasonable, too!

**Comment by BenjaminW3** [comment]
```--x cu``` seemed to be the best solution but after some tests I think the CMake FindCUDA script does not take this flag into account.
FindCUDA is filtering the files itself before calling nvcc, so there is no chance for the flag to have any effect [see here](https://github.com/Kitware/CMake/blob/9b2f69925a59ac5bafdfafb650a2f2f785348085/Modules/FindCUDA.cmake#L1320).
We have to set the ```CUDA_SOURCE_PROPERTY_FORMAT``` per file to bring FindCUDA to add ```--x cu``` for us.

**Comment by ax3l** [comment]
k, be aware that `CUDA_SOURCE_PROPERTY_FORMAT` needs [cmake 3.3.0 or newer](https://github.com/Kitware/CMake/commit/8313de2d5adf4514f0b3cccca747e78ca07d3ad3#diff-4740178b5defb85b802090e7f04ae7e2)

**Comment by BenjaminW3** [comment]
That is true...
In my test branch it works and allows to write code fully in cpp files.
This is a great usability enhancement. Is this enough to justify a raise of the minimal CMake version to 3.3.0?

**Comment by BenjaminW3** [comment]
I have implemented it for the alpaka examples in branch `topic-alpaka_add_executable`.

**Comment by erikzenker** [comment]
:1:

**Comment by ax3l** [comment]
> That is true...
> In my test branch it works and allows to write code fully in cpp files.
> This is a great usability enhancement. Is this enough to justify a raise
> of the minimal CMake version to 3.3.0?

Sure, go for it.

Pls don't forget to add it in the required version (first line of
CMakeLists.txt) and install instructions.

**Comment by BenjaminW3** [comment]
Merged into develop as of 40a846506d0d24d7c723b127ca1cab5869c83ecd.

**Comment by ax3l** [comment]
great reduction of overhead, thx!
